### DISCUSSION TITLE
Support for much faster LAPACK routine for SVD with MKL provider

### DISCUSSION THREAD

**Discussion by Adam-Re** [mathnet/mathnet-numerics#914] [Ideas]
I have noticed that with MKL provider (but probably also other providers which make use of LAPACK) the SVD function uses gesvd routine from LAPACK. LAPACK has also routine gesdd which is much more efficient for large matrices (see tables 3.18 a 3.19 [here](https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.netlib.org%2Flapack%2Flug%2Fnode71.html&data=04%7C01%7CARemo2%40imf.org%7Cd18fca324c0443ae22d708da109be9c4%7C8085fa43302e45bdb171a6648c3b6be7%7C0%7C0%7C637840559081650738%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=rGOyIOaMdbP5Kin1T82QB6aw%2Bj5H8pK5bs5NUPd5H4s%3D&reserved=0)).

Would it be possible to modify SVD to use optionally gesdd instead of gesvd?

**Comment by cdrnet** [comment]
Thanks! Would there be an argument against switching entirely to gsedd instead of gsevd - instead of providing both options?

**Reply by Adam-Re** [reply]
Personally, I haven't had any issues using gesdd instead of gesvd but I have seen discussions (e.g., https://savannah.gnu.org/bugs/?55564 ) where people mention that gesvd may give more accurate results at the order close to machine epsilon for poorly scaled matrices.

To me the speed is more important than the accuracy at the orders where they may differ but maybe there are other applications where such a degree of accuracy is important.

Based on LAPACK documentation of the two routines, gesdd requires larger workspace (O(min(m,n)^2)) whereas gesvd requires only O(max(m,n)) so there is a trade off between speed and memory.

**Comment by Adam-Re** [comment]
Hi there,

Do you plan to switch or add gsedd? I'm working on a project and we soon will need to decide whether we will use mathnet or make our own implementation based on gesdd. The difference in performance of the two routines is a deal breaker for us.
